lib/request-execution.js

```diff
@@ -193,22 +193,38 @@ class RequestExecution {
       // No point in retrying
       return;
     }
+
     this._parent.log('info', 'Retrying request');
     this._retryCount++;
+
     if (typeof consistency === 'number' && this._request.consistency !== consistency) {
       this._request = this._request.clone();
       this._request.consistency = consistency;
     }
+
     if (useCurrentHost !== false) {
-      // Use existing host (default)
-      return this._sendOnConnection();
+      // Use existing host (default).
+      // Reusing the existing connection is suitable for the most common scenarios, like server read timeouts that
+      // will be fixed with a new request.
+      // To cover all scenarios (e.g., where a different connection to the same host might mean something different),
+      // we obtain a new connection from the host pool.
+      return this._host.borrowConnection((err, connection) => {
```

Comment:
Might this just return the same connection that was previously used?  Should we add some logic to `borrowConnection` to allow passing existing connection(s) which communicates 'give me a connection that isn't any of these'?

Reply:
I was considering that the connection was likely to be removed from the pool by then (`removeFromPool()` is called on `_handleError()` method), but I haven't consider that the retry might be triggered by errors that don't cause the connection to be removed (e.g., `overloaded`).

Great remark! I'll change it to try to avoid using the same connection when possible (more than 1 open connection):
- When there's 2 or more connections, use another one.
- When there's only 1 connection, use the same one.

Reply:
This could be a good optimization for the other drivers as well.  The current state of this PR behaves exactly like the java driver.  Maybe a good confab discussion.